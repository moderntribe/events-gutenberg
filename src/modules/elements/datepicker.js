/**
 * External dependencies
 */
import moment from 'moment';
import { times, unescape, get, escapeRegExp } from 'lodash';
import { stringify } from 'querystringify';

/**
 * WordPress dependencies
 */
import { Component, compose } from '@wordpress/element';
import {
	Placeholder,
	Spinner,
	DatePicker as WPDatePicker,
	Dropdown,
	withAPIData,
} from '@wordpress/components';
import { query } from '@wordpress/data'
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */


/**
 * Module Code
 */

// Fetches all the Editor Settings
const DATA = tribe_blocks_editor_settings;

function round15( minute ) {
	let intervals = [ 15, 30, 45, 59, 0 ];
	let closest;
	let min = 90;

	for ( let i = 0; i < intervals.length; i++ ) {
		let iv = intervals[ i ];
		let maybeMin = Math.abs( minute - iv );

		if ( maybeMin < min ) {
			min = maybeMin;
			closest = iv;
		}
	}

	if ( 59 === closest ) {
		closest = 0;
	}

	return closest;
}

const toMomentFormat = ( format ) => {
	const strtr = ( str, pairs ) => {
		const substrs = Object.keys( pairs ).map( escapeRegExp )
		return str.split( RegExp( `(${ substrs.join( '|' ) })` ) )
			.map( part => pairs[ part ] || part )
			.join( '' )
	}

    const replacements = {
        'd': 'DD',
        'D': 'ddd',
        'j': 'D',
        'l': 'dddd',
        'N': 'E',
        'S': 'o',
        'w': 'e',
        'z': 'DDD',
        'W': 'W',
        'F': 'MMMM',
        'm': 'MM',
        'M': 'MMM',
        'n': 'M',
        't': '', // no equivalent
        'L': '', // no equivalent
        'o': 'YYYY',
        'Y': 'YYYY',
        'y': 'YY',
        'a': 'a',
        'A': 'A',
        'B': '', // no equivalent
        'g': 'h',
        'G': 'H',
        'h': 'hh',
        'H': 'HH',
        'i': 'mm',
        's': 'ss',
        'u': 'SSS',
        'e': 'zz', // deprecated since version 1.6.0 of moment.js
        'I': '', // no equivalent
        'O': '', // no equivalent
        'P': '', // no equivalent
        'T': '', // no equivalent
        'Z': '', // no equivalent
        'c': '', // no equivalent
        'r': '', // no equivalent
        'U': 'X',
    };

    return strtr( format, replacements );
}

class DatePicker extends Component {
	constructor() {
		super( ...arguments );

		this.onChange = this.onChange.bind( this );
		this.changeDatetime = this.props.changeDatetime.bind( this );
		this.formatDate = this.formatDate.bind( this );

		this.onChange( this.props.datetime );
	}

	formatDate( date, label = false ) {
		// Defaults to the non-label format
		let format = 'YYYY-MM-DD HH:mm:00';

		if ( ! date ) {
			date = moment();
		}

		// Convert to moment
		if ( 'string' === typeof date ) {
			date = moment( date );
		}

		// on Invalid date we reset to now
		if ( ! date.isValid() ) {
			date = moment();
		}

		if ( label ) {
			format = `${ toMomentFormat( DATA.dateWithYearFormat || 'F j, Y' ) }`
		}

		return date.minutes( round15( date.minutes() ) ).format( format );
	}

	onChange( date ) {
		// Start defaults to now
		let dateMoment = moment();

		// if we have the date we prep the moment
		if ( date ) {
			dateMoment = moment( date );
		}

		this.changeDatetime( this.formatDate( dateMoment ) );
	}

	render() {
		return (
			<Dropdown
				className="tribe-editor-datetime"
				position="bottom left"
				contentClassName="tribe-editor-datetime__dialog"
				renderToggle={ ( { onToggle, isOpen } ) => (
					<button
						type="button"
						className="button-link"
						onClick={ onToggle }
						aria-expanded={ isOpen }
					>
						{ this.formatDate( this.props.datetime, true ) }
					</button>
				) }
				renderContent={ () => <WPDatePicker key="date-picker" currentDate={ this.props.datetime } onChange={ this.onChange } /> }
			/>
		);
	}
}

export default DatePicker;
