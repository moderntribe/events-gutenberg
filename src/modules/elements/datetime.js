/**
 * External dependencies
 */
import moment from 'moment';
import { connect } from 'react-redux';
import { times, unescape, get } from 'lodash';
import { stringify } from 'querystringify';

/**
 * WordPress dependencies
 */
import { Component, compose } from '@wordpress/element';
import {
	Placeholder,
	Spinner,
	DateTimePicker,
	Dropdown,
	withAPIData,
} from '@wordpress/components';
import { query } from '@wordpress/data'
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */


/**
 * Module Code
 */
function round15( minute ) {
	let intervals = [ 15, 30, 45, 59, 0 ];
	let closest;
	let min = 90;

	for ( let i = 0; i < intervals.length; i++ ) {
		let iv = intervals[ i ];
		let maybeMin = Math.abs( minute - iv );

		if ( maybeMin < min ) {
			min = maybeMin;
			closest = iv;
		}
	}

	if ( 59 === closest ) {
		closest = 0;
	}

	return closest;
}

class DateTime extends Component {
	constructor() {
		super( ...arguments );

		this.onChange = this.onChange.bind( this );
		this.changeDatetime = this.props.changeDatetime.bind( this );
		this.formatDate = this.formatDate.bind( this );

		this.onChange( this.props.datetime );
	}

	formatDate( date, label = false ) {
		const format = label ? 'YYYY-MM-DD @ HH:mm' : 'YYYY-MM-DD HH:mm:00';

		if ( ! date ) {
			date = moment();
		}

		// Convert to moment
		if ( 'string' === typeof date ) {
			date = moment( date );
		}

		return date.minutes( round15( date.minutes() ) ).format( format );
	}

	onChange( date ) {
		// Start defaults to now
		var dateMoment = moment();

		// if we have the date we prep the moment
		if ( date ) {
			dateMoment = moment( date );
		}

		this.changeDatetime( this.formatDate( dateMoment ) );
	}

	render() {
		const dropdown = (
			<Dropdown
				className="tribe-editor-datetime"
				position="bottom left"
				contentClassName="tribe-editor-datetime__dialog"
				renderToggle={ ( { onToggle, isOpen } ) => (
					<button
						type="button"
						className="button-link"
						onClick={ onToggle }
						aria-expanded={ isOpen }
					>
						{ this.formatDate( this.props.datetime, true ) }
					</button>
				) }
				renderContent={ () => <DateTimePicker key="datetime-picker" currentDate={ this.props.datetime } onChange={ this.onChange } /> }
			/>
		);

		const label = this.props.label ? <strong className='tribe-detail-label'>{ this.props.label } </strong> : '';

		return [
			<div
				key='tribe-element-datetime'
				className='tribe-element tribe-element__datetime'
			>
				{ label }
				{ dropdown }
			</div>
		];
	}
}

export default DateTime;
