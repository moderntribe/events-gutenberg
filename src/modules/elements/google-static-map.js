/**
 * External dependencies
 */
import { stringify } from 'querystringify';
import { values } from 'lodash'
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

import { Component } from '@wordpress/element';

import {
	Placeholder,
} from '@wordpress/components';

/**
 * Internal dependencies
 */


/**
 * Module Code
 */
const IMAGE_FORMATS = {
	// png8 or png (default) specifies the 8-bit PNG format.
	PNG: 'png',

	// png32 specifies the 32-bit PNG format.
	PNG32: 'png32',

	// gif specifies the GIF format.
	GIF: 'gif',

	// jpg specifies the JPEG compression format.
	JPG: 'jpg',

	// jpg-baseline specifies a non-progressive JPEG compression format.
	JPG_BASELINE: 'jpg-baseline',
};

const MAP_TYPES = {
	// roadmap (default) specifies a standard roadmap image, as is normally shown on the Google Maps website.
	ROADMAP: 'roadmap',

	// satellite specifies a satellite image.
	SATELLITE: 'satellite',

	// terrain specifies a physical relief map image, showing terrain and vegetation.
	TERRAIN: 'terrain',

	// hybrid specifies a hybrid of the satellite and roadmap image,
	// showing a transparent layer of major streets and place names on the satellite image.
	HYBRID: 'hybrid',
};

const IMAGE_FORMATS_VALUES = values( IMAGE_FORMATS );
const MAP_TYPES_VALUES = values( MAP_TYPES );

/**
 * A wrapper for Google's Static Maps
 *
 * @see https://developers.google.com/maps/documentation/staticmaps/intro#Overview
 *
 * @example: http://staticmapmaker.com/google/
 */
class GoogleStaticMap extends Component {

	/**
	 * https://developers.google.com/maps/documentation/staticmaps/intro#api_key
	 */
	static ApiKey = null

	static RootUrl = 'http://maps.googleapis.com/maps/api/staticmap';

	static ImageFormats = IMAGE_FORMATS

	static MapTypes = MAP_TYPES

	static propTypes = {
		latitude: PropTypes.string.isRequired,

		longitude: PropTypes.string.isRequired,

		size: PropTypes.shape( {
			width: PropTypes.number.isRequired,
			height: PropTypes.number.isRequired,
		} ),

		style: PropTypes.object,

		/**
		 * zoom (required if markers not present) defines the zoom level of the map,
		 * which determines the magnification level of the map.
		 *
		 * @see https://developers.google.com/maps/documentation/staticmaps/intro#Zoomlevels
		 */
		zoom: PropTypes.number.isRequired,

		/**
		 * scale affects the number of pixels that are returned.
		 * scale=2 returns twice as many pixels as scale=1 while retaining the same coverage area and level of detail
		 * The default value is calculated from the screen PixelRatio.
		 */
		scale: PropTypes.number,

		/**
		 * @see https://developers.google.com/maps/documentation/staticmaps/intro#ImageFormats
		 */
		format: PropTypes.oneOf( IMAGE_FORMATS_VALUES ),

		/**
		 * @see https://developers.google.com/maps/documentation/staticmaps/intro#MapTypes
		 */
		mapType: PropTypes.oneOf( MAP_TYPES_VALUES ),

		/**
		 * Add a marker on the center
		 */
		hasCenterMarker: PropTypes.bool,
	}

	static defaultProps = {
		format: IMAGE_FORMATS.JPG,
		mapType: MAP_TYPES.ROADMAP,
		hasCenterMarker: true,
		style: {},
		scale: 2,
	}

	render() {
		let mapElement = (
			<Placeholder style={{ height: '100%' }}>
				<p>{ __( 'No map preview available', 'the-events-calendar' ) }</p>
			</Placeholder>
		)

		const {
			apiKey,
		} = this.props;

		if ( apiKey ) {
			mapElement = (
				<img
					className='tribe-editor-map__element'
					src={ this.staticMapUrl }
				/>
			)
		}

		return (
			<div className='tribe-editor-map__container'>
				{ mapElement }
			</div>
		);
	}

	get staticMapUrl() {
		const {
			latitude,
			longitude,
			zoom,
			size,
			scale,
			format,
			mapType,
			apiKey,
		} = this.props;

		const { width, height } = size;
		const rootUrl = this.constructor.RootUrl;

		const queryArgs = {
			center: `${latitude},${longitude}`,
			zoom: zoom,
			scale: scale,
			size: `${width}x${height}`,
			maptype: mapType,
			format: format,
			markers: this.markerParams,
			key: apiKey,
		}

		return `${rootUrl}?${ stringify( queryArgs ) }`;
	}

	get markerParams() {
		const {
			latitude,
			longitude,
			hasCenterMarker,
		} = this.props;

		const markerParams = `size:mid|color:0xff0000|label:|${latitude},${longitude}`;
		return hasCenterMarker ? markerParams : '';
	}
}

export default GoogleStaticMap;