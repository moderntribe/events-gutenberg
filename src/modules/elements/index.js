export { default as Input } from './input/element.js';
export { default as TermsList } from './terms-list/element.js';
export { default as OrganizerForm } from './organizer-form/element.js';
export { default as VenueForm, toFields, toVenue } from './venue-form/element.js';
export { default as DatePicker } from './date-picker/element.js';
export { default as TimePicker } from './time-picker/element.js';
export { default as MetaGroup } from './meta-group/element.js';
export { default as SearchPosts } from './search-posts/element.js';
export { default as GoogleMap } from './google-map/element.js';
export { default as Dashboard } from './dashboard/element';
export { default as CheckBox } from './checkbox/element';
export { default as Month } from './month/element';
export { default as SearchOrCreate } from './search-or-create/element';
export { default as Loading } from './loading';
