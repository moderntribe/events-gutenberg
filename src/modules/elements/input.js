const {Component} = wp.element
import PropTypes from 'prop-types'
import validator from 'validator';
import { isFunction, noop } from 'lodash';

class Input extends Component {

	static propTypes = {
		validated: PropTypes.bool,
		required: PropTypes.bool,
	}

	static defaultProps = {
		required: false,
		validate: false,
	}

	constructor() {
		super(...arguments);

		this.state = {
			isValid: this.validate(''),
		}
	}

	onChange = (event) => {
		const { onChange, onComplete, validate } = this.props;
		let callback = isFunction( onChange ) ? onChange : noop;
		let completeCallback = isFunction( onComplete ) ? onComplete : noop;

		if ( validate ) {
			this.setState({ isValid: this.validate(event.target.value) }, completeCallback );
			callback(event);
		} else {
			completeCallback();
			callback(event);
		}
	}

	validate(value) {
		const { validateCallback } = this.props;
		return isFunction( validateCallback ) ? validateCallback(value) : this.maybeValidate(value);
	}

	maybeValidate = ( value ) => {
		const { type, required } = this.props;

		if ( value.length === 0 ) {
			return ! required;
		}

		let isValid = true;
		switch (type) {
			case 'phone':
				isValid = validator.isMobilePhone( value, 'any' );
			break;
			case 'email':
				isValid = validator.isEmail( value );
				break;
			case 'url':
				isValid = validator.isURL( value );
				break;
		}
		return isValid;
	}

	isValid() {
		return this.state.isValid;
	}

	focus() {
		this.input.focus();
	}

	getClassName() {
		const { className, validate } = this.props;
		const { isValid } = this.state;
		let classes = className ? className.split( ' ' ) : [];

		if ( validate ) {
			if ( isValid ) {
				classes.push( 'is-valid' );
			} else {
				classes.push( 'is-invalid' );
			}
		}

		return classes
			.filter( (className) => className && className.length )
			.join(' ');
	}

	render () {
		// Remove properties that are not part of the DOM.
		const { onComplete, required, validate, ...properties} = this.props;
		return <input
			{...properties}
			className={`${this.getClassName()}`}
			ref={(input) => this.input = input }
			onChange={this.onChange}
		/>
	}
}

export default Input