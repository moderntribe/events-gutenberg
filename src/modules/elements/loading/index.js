/**
 * External dependencies
 */
import classNames from 'classnames';

/**
 * Wordpress dependencies
 */
import { Spinner } from '@wordpress/components';

import './style.pcss';

export default ( { className } ) => {
	return (
		<span className={ classNames( [ "tribe-editor__spinner-container", className ] ) }>
			<Spinner />
		</span>
	);
};
