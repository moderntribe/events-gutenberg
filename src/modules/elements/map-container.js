/**
 * External dependencies
 */
import { withScriptjs, withGoogleMap, GoogleMap, Marker } from 'react-google-maps'
import { geocodeByAddress, getLatLng } from 'react-places-autocomplete'

/**
 * WordPress dependencies
 */
import { query } from '@wordpress/data'
import { Component, compose } from '@wordpress/element';

import {
	Dropdown,
	IconButton,
	Dashicon,
	Spinner,
	Placeholder,
	withAPIData,
	withState
} from '@wordpress/components';

/**
 * Internal dependencies
 */

/**
 * Module Code
 */
class MapContainer extends Component {
	static defaultProps = {
		key: 'AIzaSyDquv6yMOGPHM6-Y7MkI5gv20mCS740M8A',
		center: null,
		zoom: 14,
		address: null,
	};

	componentDidMount() {
		this.props.onRef( this )
	}

	componentWillUnmount() {
		this.props.onRef( undefined )
	}

	constructor() {
		super( ...arguments );

		this.state = this.props

		this.updateAddress = this.updateAddress.bind( this )
		this.updateAddress( this.props.address )
	}

	updateAddress( address ) {
		if ( ! address ) {
			this.setState( { center: null, address: null } )
		}

		geocodeByAddress( address )
			.then( results => getLatLng( results[0] ) )
			.then( latLng => {
				this.setState( { center: latLng, address: address } )
				console.debug( 'Successfully Geocoded Address', latLng, address )
			} )
			.catch( error => {
				console.debug( 'Error on Geocoding Address', error, address )
			} )
	}

	render() {
		const { zoom, center } = this.state;

		if ( ! center ) {
			return null
		}

		return (
			<GoogleMap
				defaultZoom={ zoom }
				defaultCenter={ center }
			>
				<Marker
					position={ center }
				/>
			</GoogleMap>
		);
	}
}

export default compose(
	withState( {
		googleMapURL: 'https://maps.googleapis.com/maps/api/js?key=AIzaSyDquv6yMOGPHM6-Y7MkI5gv20mCS740M8A&v=3.exp&libraries=geometry,drawing,places',
		isMarkerShown: false,
		loadingElement: ( <Placeholder key="placeholder"><Spinner /></Placeholder> ),
		containerElement: ( <div className='tribe-editor-map__container' /> ),
		mapElement: ( <div className='tribe-editor-map__element' /> ),
	} ),
	withScriptjs,
	withGoogleMap,
)( MapContainer )