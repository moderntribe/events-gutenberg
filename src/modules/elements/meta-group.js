/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */
import { Component } from '@wordpress/element';

/**
 * Internal dependencies
 */

/**
 * Module Code
 */
class MetaGroup extends Component {
	static defaultProps = { className: '', children: null }

	constructor() {
		super( ...arguments );
	}

	render() {
		const { groupKey, className, children } = this.props;

		return (
			<div
				className={ `tribe-editor-meta-group tribe-editor-meta-group-${groupKey} ${className}` }
				key={ groupKey }
			>
				{ children }
			</div>
		)
	}
}
