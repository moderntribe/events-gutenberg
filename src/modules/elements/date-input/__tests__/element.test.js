/**
 * External dependencies
 */
import renderer from 'react-test-renderer';
import React from 'react';

/**
 * Internal dependencies
 */
import { DateInput } from 'elements';

describe( 'DateInput element', () => {
	test( 'Should render the component', () => {
		const clickMock = jest.fn();
		const setDateMock = jest.fn();
		const component = renderer.create(
			<DateInput
				selected={ true }
				onClickHandler={ clickMock }
				setDateTime={ setDateMock }
			>
				<div>This is the children</div>
			</DateInput>,
		);
		expect( component.toJSON() ).toMatchSnapshot();
	} );
} );
