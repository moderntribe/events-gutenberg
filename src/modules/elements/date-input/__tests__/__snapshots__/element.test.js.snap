// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateInput element Should render the component 1`] = `
<div
  onClick={[MockFunction]}
  onKeyDown={[MockFunction]}
  role="button"
  tabIndex={0}
>
  <input
    className="tribe-editor__date-input"
    name="date-input"
    onBlur={[Function]}
    placeholder="Enter your date"
    type="text"
  />
</div>
`;
